lib/syslog/writer.go

```diff
@@ -244,6 +300,50 @@ func (c bufferedConn) Close() error                { return c.conn.Close() }
 func (c bufferedConn) Flush() error                { return c.buf.Flush() }
 func (c bufferedConn) Write(b []byte) (int, error) { return c.buf.Write(b) }
 
+func cachedDialWriter(network string, address string, config *tls.Config, socksProxy string) (w io.Writer, err error) {
+	if w, ok := writerCache.get(network, address, socksProxy); ok {
+		return w, nil
+	}
+	w, err = syncedDialWriter(network, address, config, socksProxy)
+	if err != nil {
+		writerCache.set(network, address, socksProxy, w)
+	}
+	return w, err
+}
+
+type dialerMutexMap struct {
+	sync.Mutex
+	mu map[string]*sync.Mutex
+}
+
+func (mm dialerMutexMap) lock(network string, address string, socksProxy string) {
+	key := writerCacheKey(network, address, socksProxy)
```

Comment:
How about doing the key computation outside of the method?
1. That way it doesn't get recomputed for lock and unlock
2. it makes your data structure more generic if it doesn't make assumptions about the context in which it's used